### ISSUE TITLE
CoordinateSequence-friendly MultiPoint

### ISSUE THREAD

**Issue by jagill** [locationtech/jts#625]
Currently, MultiPoints contain a List of Points.  If they supported CoordinateSequence for the internal storage, we could use the more efficient PackedCoordinateSequence.

**Comment by juliusfriedman** [comment]
Geometries are meant to be thought of as primitive types.
See also https://github.com/NetTopologySuite/NetTopologySuite/issues/451

Perhaps

```class EfficientGeometry: GeometryCollection{ CoordinateSequence ephermal; Geometry flattened; } ```

```class EfficientMultiPoint : EfficientGeometry{ MultiPoint flattened; } ```

**Comment by dr-jts** [comment]
> If they supported CoordinateSequence for the internal storage, we could use the more efficient PackedCoordinateSequence.

Yes, but at the price of a more expensive `getGeometryN` call, since it would have to create a new `Point` each time.

Not sure which is better - depends on use case, I guess.  There are lots of JTS algorithms which iterate over every element of a geometry collection, though.  But if the `MultiPoint` is being created just to _possibly_ run **one** eventual operation, then it is better to defer the cost of Point creation.

Ideally JTS should provide all geometry types as interfaces, and then the implementation could be chosen based on need.  This is absolutely on the roadmap for JTS 2, but it's a big effort to do in JTS 1.

**Comment by FObermaier** [comment]
MultiPoint and its Points could live with one main sequence and a type of sequence that points to parts of that main sequence:
A proto-type that requires some additions to jts-core:
https://github.com/locationtech/jts/compare/master...FObermaier:lab/sequence_backed_coordinate

**Comment by dr-jts** [comment]
> MultiPoint and its Points could live with one main sequence and a type of sequence that points to parts of that main sequence:

Not sure what this is trying to address?  It would still require creation of `Coordinate` objects, right?  They will be somewhat smaller than a regular `Coordinate`. I'm not sure that matters that much in the context of most algorithms.  

This also complicates the `Coordinate` model even more.  The current provision for a few different Coordinate subclasses is already causing headaches, so proliferating this will only make the problem worse.
